Add com.github.inercia.k3x

inercia opened the pull request:
Initial commit for [k3x](https://github.com/inercia/k3x), a graphical user interface for [k3d](https://github.com/rancher/k3d), making it trivial to have your own local Kubernetes cluster(s) in your laptop.

LeandroStanger commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,124 @@
+{
+  "app-id": "com.github.inercia.k3x",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.34",
```
Update runtime to 3.36
```suggestion
  "runtime-version": "3.36",
```

LeandroStanger commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,124 @@
+    "--talk-name=org.kde.StatusNotifierWatcher",
+    "--talk-name=com.canonical.indicator.application",
+    "--talk-name=org.ayatana.indicator.application",
+    "--talk-name=ca.desrt.dconf",
+    "--filesystem=xdg-run/dconf",
+    "--filesystem=~/.config/dconf:ro",
+    "--env=DCONF_USER_CONFIG_DIR=.config/dconf",
```
```suggestion
```

barthalion commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+    "--socket=wayland",
+    "--socket=ssh-auth",
+    "--share=network",
+    "--env=LD_LIBRARY_PATH=/app/lib:/app/lib/x86_64-linux-gnu:/lib64:/lib:/usr/lib64:/usr/lib:/usr/local/lib",
```
I don't think this is needed.

bilelmoussaoui replied:
Actually per https://github.com/elementary/granite/issues/241 it's the only way for us to make it work

barthalion commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+    "--talk-name=org.kde.StatusNotifierWatcher",
+    "--talk-name=com.canonical.indicator.application",
+    "--talk-name=org.ayatana.indicator.application",
+    "--filesystem=home:rw"
```
Does it need full RW access or are there specific directories it uses?

inercia replied:
It could probably be fine tunned, but users can set a [_hook script_ that will be executed when the cluster is created/destroyed](https://github.com/inercia/k3x/blob/master/docs/user-manual-preferences.md#scripts). So that means k3x should be able to open and run those files. In addition, it needs to access:

- `~/.kube` for writing/readfing `kubeconfig` files
- `/.config/autostart` for adding k3x to the _autostart_ list

barthalion commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+      "buildsystem": "simple",
+      "build-options": {
+        "build-args": [
+          "--share=network"
```
Flathub builds are offline to ensure reproducibility; you can generate JSON snippet for these libraries with https://github.com/flatpak/flatpak-builder-tools/tree/master/pip.

barthalion commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+    },
+    {
+      "name": "k3x",
+      "builddir": true,
```
IIRC `builddir` is set by default for meson.

barthalion commented:
bot, build com.github.inercia.k3x

barthalion commented:
Also looks like you need to add shared-modules git submodule.

bilelmoussaoui commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+    {
+      "name": "libappindicator"
+    },
+    {
+      "name": "granite",
+      "buildsystem": "meson",
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/elementary/granite.git",
+          "tag": "5.3.1"
+        }
+      ]
+    },
+    {
+      "name": "elementary-icons",
+      "buildsystem": "meson",
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/elementary/icons.git",
+          "tag": "5.0"
+        }
+      ]
+    },
+    {
```
Any reason for not using `io.elementary.BaseApp`?

inercia replied:
Good suggestion. I will change that. Thanks

bilelmoussaoui commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+  ],
+  "modules": [
+    "shared-modules/libappindicator/libappindicator-gtk3-introspection-12.10.json",
+    {
+      "name": "libappindicator"
+    },
```
```suggestion
```
seems to be an empty module

bilelmoussaoui commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+{
+  "app-id": "com.github.inercia.k3x",
+  "runtime": "org.gnome.Platform",
```
Does it really need the GNOME Sdk?

inercia replied:
Replaced by:
```json
  "base": "io.elementary.BaseApp",
  "base-version": "juno-19.08",
  "runtime": "org.freedesktop.Platform",
  "sdk": "org.freedesktop.Sdk",
  "runtime-version": "19.08",
```

bilelmoussaoui commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,120 @@
+  "finish-args": [
+    "--device=dri",
+    "--share=ipc",
+    "--socket=x11",
```
```suggestion
    "--socket=fallback-x11",
```

bilelmoussaoui commented:
bot, build com.github.inercia.k3x

barthalion commented:
bot, build com.github.inercia.k3x

barthalion commented:
Not sure what this failure is about to be honest…

barthalion commented:
Looks like this confuses flatpak-builder:

https://github.com/inercia/k3x/commit/e882d919898e1a724c4c13c559d98f5d4d2d3cce#diff-8903239df476d7401cf9e76af0252622R4

inercia commented:
> Looks like this confuses flatpak-builder: inercia/k3x@e882d91#diff-8903239df476d7401cf9e76af0252622R4

Thanks @barthalion. I have removed that submodule...

bilelmoussaoui commented:
bot, build com.github.inercia.k3x

inercia commented:
It seems this configuration:

```json
  "runtime": "org.freedesktop.Platform",
  "runtime-version": "19.08",
  "sdk": "org.freedesktop.Sdk",
  "base": "io.elementary.BaseApp",
  "base-version": "juno-19.08",
```

does not work:

```
Traceback (most recent call last):
  File "/app/bin/k3x", line 41, in <module>
    import gi
ModuleNotFoundError: No module named 'gi'
Makefile:157: recipe for target 'run' failed
make: *** [run] Error 1
```

And the build is failing because of the `appdata`: I will fix it.

bilelmoussaoui commented:
> It seems this configuration:
> 
> ```json
>   "runtime": "org.freedesktop.Platform",
>   "runtime-version": "19.08",
>   "sdk": "org.freedesktop.Sdk",
>   "base": "io.elementary.BaseApp",
>   "base-version": "juno-19.08",
> ```
> 
> does not work when trying to run `k3x`:
> 
> ```
> Traceback (most recent call last):
>   File "/app/bin/k3x", line 41, in <module>
>     import gi
> ModuleNotFoundError: No module named 'gi'
> ```
> 
> And the build is failing because of the `appdata`: I will fix it.

Using freedesktop runtime isn't a requirement, feel free to switch to GNOME runtime which includes the missing dependency.

bilelmoussaoui commented:
bot, build com.github.inercia.k3x

inercia commented:
IKt does not work yet:

```
$ flatpak run com.github.inercia.k3x                                                              
GI search path:
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0/
[PRE-MAIN]  [search-path] - /usr/lib/x86_64-linux-gnu/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN] Current LD_LIBRARY_PATH: 

(k3x:2): Gtk-WARNING **: 18:10:24.732: Locale not supported by C library.
	Using the fallback 'C' locale.
Gtk-Message: 18:10:24.732: Failed to load module "pantheon-filechooser-module"

** (k3x:2): WARNING **: 18:10:24.791: Failed to load shared library 'libgranite.so.5.3.0' referenced by the typelib: libgranite.so.5.3.0: cannot open shared object file: No such file or directory
/usr/lib/python3.7/site-packages/gi/types.py:220: Warning: cannot derive 'k3x+utils_ui+SettingsPage' from non-derivable parent type 'void'
  _gi.type_register(cls, namespace.get('__gtype_name__'))
Traceback (most recent call last):
  File "/app/bin/k3x", line 68, in <module>
    from k3x import main
  File "/app/share/k3x/k3x/main.py", line 43, in <module>
    from .menu import K3dvMenu
  File "/app/share/k3x/k3x/menu.py", line 29, in <module>
    from .cluster_view import ClusterDialog
  File "/app/share/k3x/k3x/cluster_view.py", line 45, in <module>
    from .k3d import K3dCluster
  File "/app/share/k3x/k3x/k3d.py", line 42, in <module>
    from .utils_ui import show_notification
  File "/app/share/k3x/k3x/utils_ui.py", line 256, in <module>
    class SettingsPage(Granite.SimpleSettingsPage):
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 229, in __init__
    super(GObjectMeta, cls).__init__(name, bases, dict_)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 208, in __init__
    cls._type_register(cls.__dict__)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 220, in _type_register
    _gi.type_register(cls, namespace.get('__gtype_name__'))
RuntimeError: could not create new GType: k3x+utils_ui+SettingsPage (subclass of void)
```

inercia commented:
I had to [add `'/usr/lib/x86_64-linux-gnu/girepository-1.0'` to the GI search path](https://github.com/inercia/k3x/blob/master/src/k3x.in#L53), not sure why this is necessary...

barthalion commented:
bot, build com.github.inercia.k3x

barthalion commented:
It still does not start for me:

```
GI search path:
[PRE-MAIN]  [search-path] - /usr/lib/x86_64-linux-gnu/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN] Current LD_LIBRARY_PATH: 

** (k3x:2): WARNING **: 11:10:32.079: Failed to load shared library 'libgranite.so.5.3.0' referenced by the typelib: libgranite.so.5.3.0: cannot open shared object file: No such file or directory
/usr/lib/python3.7/site-packages/gi/types.py:220: Warning: cannot derive 'k3x+utils_ui+SettingsPage' from non-derivable parent type 'void'
  _gi.type_register(cls, namespace.get('__gtype_name__'))
Traceback (most recent call last):
  File "/app/bin/k3x", line 61, in <module>
    from k3x import main
  File "/app/share/k3x/k3x/main.py", line 43, in <module>
    from .menu import K3dvMenu
  File "/app/share/k3x/k3x/menu.py", line 29, in <module>
    from .cluster_view import ClusterDialog
  File "/app/share/k3x/k3x/cluster_view.py", line 45, in <module>
    from .k3d import K3dCluster
  File "/app/share/k3x/k3x/k3d.py", line 42, in <module>
    from .utils_ui import show_notification
  File "/app/share/k3x/k3x/utils_ui.py", line 256, in <module>
    class SettingsPage(Granite.SimpleSettingsPage):
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 229, in __init__
    super(GObjectMeta, cls).__init__(name, bases, dict_)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 208, in __init__
    cls._type_register(cls.__dict__)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 220, in _type_register
    _gi.type_register(cls, namespace.get('__gtype_name__'))
RuntimeError: could not create new GType: k3x+utils_ui+SettingsPage (subclass of void
```

The file is here though:

```
$ flatpak run --command=find com.github.inercia.k3x /app/lib -name libgranite.so.5.3.0
/app/lib/libgranite.so.5.3.0
```

Similarly I'm pretty sure `/usr/lib/x86_64-linux-gnu/girepository-1.0` is in the default path unless it's overwritten somewhere, which may be the case given that BaseApp bundles it's own GI: https://github.com/flathub/io.elementary.BaseApp/blob/branch/juno-19.08/io.elementary.BaseApp.yml#L25

Is it something you can take a look @bilelmoussaoui?

barthalion commented:
@inercia: Desktop file uses non-existent binary in `Exec=`. Also in the case when Docker socket is not available, it would be nice to display a graphical error instead of traceback:

```
Traceback (most recent call last):
  File "/app/lib/python3.7/site-packages/urllib3/connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "/app/lib/python3.7/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.7/http/client.py", line 1244, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/lib/python3.7/http/client.py", line 966, in send
    self.connect()
  File "/app/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
    sock.connect(self.unix_socket)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/app/lib/python3.7/site-packages/urllib3/connectionpool.py", line 725, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/app/lib/python3.7/site-packages/urllib3/util/retry.py", line 403, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/app/lib/python3.7/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/app/lib/python3.7/site-packages/urllib3/connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "/app/lib/python3.7/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.7/http/client.py", line 1244, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/lib/python3.7/http/client.py", line 966, in send
    self.connect()
  File "/app/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
    sock.connect(self.unix_socket)
urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/bin/k3x", line 62, in <module>
    sys.exit(main.main(VERSION))
  File "/app/share/k3x/k3x/main.py", line 137, in main
    _indicator = Indicator(version=version)  # NOTE: assign for keeping the object alive
  File "/app/share/k3x/k3x/main.py", line 86, in __init__
    version=self._version)
  File "/app/share/k3x/k3x/menu.py", line 104, in __init__
    self._preferences = PreferencesDialog(docker=self._docker)
  File "/app/share/k3x/k3x/preferences.py", line 132, in __init__
    self.view = PreferencesPanedView(settings=self._settings, docker=self._docker)
  File "/app/share/k3x/k3x/preferences.py", line 227, in __init__
    self.k3s_preferences = K3sSettingsPage(settings=self._settings, docker=self._docker)
  File "/app/share/k3x/k3x/preferences.py", line 432, in __init__
    for image in self._docker.get_official_k3s_images():
  File "/app/share/k3x/k3x/docker.py", line 90, in get_official_k3s_images
    return self._client.images.list("rancher/k3s")
  File "/app/lib/python3.7/site-packages/docker/models/images.py", line 363, in list
    resp = self.client.api.images(name=name, all=all, filters=filters)
  File "/app/lib/python3.7/site-packages/docker/api/image.py", line 90, in images
    res = self._result(self._get(self._url("/images/json"), params=params),
  File "/app/lib/python3.7/site-packages/docker/utils/decorators.py", line 46, in inner
    return f(self, *args, **kwargs)
  File "/app/lib/python3.7/site-packages/docker/api/client.py", line 230, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/app/lib/python3.7/site-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/app/lib/python3.7/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/app/lib/python3.7/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/app/lib/python3.7/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```

bilelmoussaoui commented:
I will have a look as soon as I can

inercia commented:
> @inercia: Desktop file uses non-existent binary in Exec=. Also in the case when Docker socket is not available, it would be nice to display a graphical error instead of traceback:

Right, I will add a check for the connectivity to the Docker daemon...

inercia commented:
>> @inercia: Desktop file uses non-existent binary in Exec=. Also in the case when Docker socket is not available, it would be nice to display a graphical error instead of traceback:

> Right, I will add a check for the connectivity to the Docker daemon...

Updated to `v0.2.4`: added a check for the Docker daemon.

nedrichards commented:
bot, build com.github.inercia.k3x

@inercia you can run that command yourself as well when you push updates.

nedrichards commented:
I get the following error running that build on Endless OS (no docker, does ship with podman)

`nick@nedr-yoga:~$ flatpak run com.github.inercia.k3x
GI search path:
[PRE-MAIN]  [search-path] - /usr/lib/x86_64-linux-gnu/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN] Current LD_LIBRARY_PATH: 

** (k3x:3): WARNING **: 20:48:21.448: Failed to load shared library 'libgranite.so.5.3.0' referenced by the typelib: libgranite.so.5.3.0: cannot open shared object file: No such file or directory
/usr/lib/python3.7/site-packages/gi/types.py:220: Warning: cannot derive 'k3x+utils_ui+SettingsPage' from non-derivable parent type 'void'
  _gi.type_register(cls, namespace.get('__gtype_name__'))
Traceback (most recent call last):
  File "/app/bin/k3x", line 61, in <module>
    from k3x import main
  File "/app/share/k3x/k3x/main.py", line 41, in <module>
    from .docker import DockerController
  File "/app/share/k3x/k3x/docker.py", line 31, in <module>
    from .utils_ui import show_notification
  File "/app/share/k3x/k3x/utils_ui.py", line 238, in <module>
    class SettingsPage(Granite.SimpleSettingsPage):
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 229, in __init__
    super(GObjectMeta, cls).__init__(name, bases, dict_)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 208, in __init__
    cls._type_register(cls.__dict__)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 220, in _type_register
    _gi.type_register(cls, namespace.get('__gtype_name__'))
RuntimeError: could not create new GType: k3x+utils_ui+SettingsPage (subclass of void)
`

inercia commented:
> I get the following error running that build on Endless OS (no docker, does ship with podman)

@nedrichards see [previous comment](https://github.com/flathub/flathub/pull/1550#issuecomment-636727159). It seems it is not finding `libgranite` from the _BaseApp_...

Besides that, k3x will note work with `podman`: it will show an error message if you get passed this problem with libgranite...

bilelmoussaoui commented:
The application requires a newer release of granite, I have updated the baseApp over here https://github.com/flathub/io.elementary.BaseApp/pull/15. I will trigger a new build once the baseApp is built & published

bilelmoussaoui commented on com.github.inercia.k3x.json:
```diff
@@ -0,0 +1,83 @@
+    "*.la",
+    "*.a"
+  ],
+  "copy-icon": true,
```
```suggestion
```
I don't think you need this? it's only useful if you have a `rename-icon`

bilelmoussaoui commented:
bot, build com.github.inercia.k3x

bilelmoussaoui commented:
bot, build com.github.inercia.k3x

inercia commented:
Still failing, although now it is `Failed to load shared library 'libgranite.so.5.4.0' ` instead of `Failed to load shared library 'libgranite.so.5.3.0'`:

```
$ flatpak run com.github.inercia.k3x                                                     
GI search path:
[PRE-MAIN]  [search-path] - /usr/lib/x86_64-linux-gnu/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN]  [search-path] - /app/lib/girepository-1.0
[PRE-MAIN] Current LD_LIBRARY_PATH: 

(k3x:2): Gtk-WARNING **: 14:57:59.096: Locale not supported by C library.
	Using the fallback 'C' locale.
Gtk-Message: 14:57:59.096: Failed to load module "pantheon-filechooser-module"

** (k3x:2): WARNING **: 14:57:59.131: Failed to load shared library 'libgranite.so.5.4.0' referenced by the typelib: libgranite.so.5.4.0: cannot open shared object file: No such file or directory
/usr/lib/python3.7/site-packages/gi/types.py:220: Warning: cannot derive 'k3x+utils_ui+SettingsPage' from non-derivable parent type 'void'
  _gi.type_register(cls, namespace.get('__gtype_name__'))
Traceback (most recent call last):
  File "/app/bin/k3x", line 61, in <module>
    from k3x import main
  File "/app/share/k3x/k3x/main.py", line 41, in <module>
    from .docker import DockerController
  File "/app/share/k3x/k3x/docker.py", line 31, in <module>
    from .utils_ui import show_notification
  File "/app/share/k3x/k3x/utils_ui.py", line 238, in <module>
    class SettingsPage(Granite.SimpleSettingsPage):
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 229, in __init__
    super(GObjectMeta, cls).__init__(name, bases, dict_)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 208, in __init__
    cls._type_register(cls.__dict__)
  File "/usr/lib/python3.7/site-packages/gi/types.py", line 220, in _type_register
    _gi.type_register(cls, namespace.get('__gtype_name__'))
RuntimeError: could not create new GType: k3x+utils_ui+SettingsPage (subclass of void)
```

Maybe I should revert to the [initial commit where `libgranite` and the icons were in the manifest](https://github.com/flathub/flathub/pull/1550/commits/e570253ed060e9ca87fbb59d0977d3de9d9be3ab#diff-1b92a5d80624efb5f9966c562413ed0aR45-R66)... :thinking:

bilelmoussaoui commented:
There's no issue with the BaseApp I guess, if you run the app with `flatpak run --env=LD_LIBRARY_PATH=/app/lib com.github.inercia.k3x` it runs, it get stuck at another level

inercia commented:
Ok, I have restored the [`--env=LD_LIBRARY_PATH=...`](https://github.com/flathub/flathub/pull/1550/files#diff-1b92a5d80624efb5f9966c562413ed0aR16).

bot, build com.github.inercia.k3x

inercia commented:
It is working now.

inercia commented:
> @inercia: Desktop file uses non-existent binary in Exec=. Also in the case when Docker socket is not available, it would be nice to display a graphical error instead of traceback:

@barthalion any suggestion on how to set the `Exec=`? It is currently defined ina template [here](https://github.com/inercia/k3x/blob/master/data/com.github.inercia.k3x.desktop.in.in#L4) and processed [here](https://github.com/inercia/k3x/blob/master/data/meson.build#L19-L42).

bilelmoussaoui commented:
> > @inercia: Desktop file uses non-existent binary in Exec=. Also in the case when Docker socket is not available, it would be nice to display a graphical error instead of traceback:
> 
> @barthalion any suggestion on how to set the `Exec=`? It is currently defined in a template [here](https://github.com/inercia/k3x/blob/master/data/com.github.inercia.k3x.desktop.in.in#L4):
> 
> ```
> Exec=@APP_ID@
> ```
> 
> and processed [here](https://github.com/inercia/k3x/blob/master/data/meson.build#L19-L42):
> 
> ```
> desktop_conf = configuration_data()
> desktop_conf.set('ICON', app_id)
> desktop_conf.set('APP_ID', app_id)
> desktop_conf.set('VERSION', meson.project_version())
> 
> desktop_file = i18n.merge_file(
>   input: configure_file(
>     input: files('@0@.desktop.in.in'.format(app_id)),
>     output: '@0@.desktop.in'.format(app_id),
>     configuration: desktop_conf
>   ),
>   output: '@0@.desktop'.format(app_id),
>   type: 'desktop',
>   po_dir: '../po',
>   install: true,
>   install_dir: join_paths(get_option('datadir'), 'applications')
> )
> 
> desktop_utils = find_program('desktop-file-validate', required: false)
> if desktop_utils.found()
>   test('Validate desktop file', desktop_utils,
>     args: [desktop_file]
>   )
> endif
> ```
> 
> Maybe just `Exec=k3x`?

The exec should be just k3x, you don't need to set it per the app-id

nedrichards commented:
bot, build com.github.inercia.k3x

barthalion commented:
/merge

inercia commented:
Thanks, @barthalion @nedrichards @bilelmoussaoui, or your reviews and your time.
